SCC check API: REST

sdminonne opened the pull request:
@soltysh

sdminonne commented:
[test]

soltysh commented on pkg/security/registry/podsecuritypolicyreview/rest.go:
```diff
@@ -0,0 +1,130 @@
+		return pspr, err
+	}
+
+	if len(serviceAccounts) == 0 {
```
I wonder if we shouldn't return an error to a user in this case. It seems reasonable.

soltysh commented on pkg/security/registry/podsecuritypolicyreview/rest.go:
```diff
@@ -0,0 +1,130 @@
+			)
+			pspsrs := securityapi.PodSecurityPolicySubjectReviewStatus{}
+			if provider, namespace, err = oscc.CreateProviderFromConstraint(ns, namespace, constraint, r.client); err != nil {
+				errs = append(errs, fmt.Errorf("unable to create provider provider for service account %s: %v", sa.Name, err))
```
unable to create provider ~~provider~~ for...

soltysh commented on pkg/security/registry/podsecuritypolicyreview/rest.go:
```diff
@@ -0,0 +1,130 @@
+		userInfo := serviceaccount.UserInfo(ns, sa.Name, "")
+		saConstraints, err := r.sccMatcher.FindApplicableSCCs(userInfo)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error finding SCC for ServiceAccount %s: %v", sa.Name, err))
```
s/error finding/unable to find

soltysh replied:
Your using full SecurityContextConstraints in other places, maybe you should do here, as well.

soltysh commented on pkg/security/registry/podsecuritypolicyreview/rest.go:
```diff
@@ -0,0 +1,130 @@
+	}
+	ns, ok := kapi.NamespaceFrom(ctx)
+	if !ok {
+		return pspr, kapierrors.NewBadRequest("namespace parameter required.")
```
You should return nil, when there was an error.

soltysh commented on pkg/security/registry/podsecuritypolicyreview/rest.go:
```diff
@@ -0,0 +1,130 @@
+	}
+	serviceAccounts, err := getServiceAccounts(pspr.Spec, r.client, ns)
+	if err != nil {
+		return pspr, err
```
`return nil, err`

soltysh commented on pkg/security/registry/podsecuritypolicyreview/rest.go:
```diff
@@ -0,0 +1,130 @@
+
+	}
+	if len(errs) > 0 {
+		return pspr, kerrors.NewAggregate(errs)
```
`return nil, kerrors.NewAggregate(errs)`

soltysh commented on pkg/security/registry/podsecuritypolicyreview/rest_test.go:
```diff
@@ -0,0 +1,96 @@
+	oscc "github.com/openshift/origin/pkg/security/scc"
+)
+
+func TestNoErrors(t *testing.T) {
```
The test cases are not covering error paths at all, see for yourself running:
`COVERAGE_OUTPUT_DIR=/tmp/origin_coverage/ hack/test-go.sh pkg/security/registry/podsecuritypolicyreview/`

soltysh replied:
At least it would be good to have:
1. invalid PSPR
2. no SA
3. failure creating provider
4. failure filling PodSecurityPolicyReviewStatus
5. PSPR with SA names in it.

soltysh commented on pkg/security/registry/podsecuritypolicyselfsubjectreview/rest.go:
```diff
@@ -0,0 +1,90 @@
+			err      error
+		)
+		if provider, namespace, err = oscc.CreateProviderFromConstraint(ns, namespace, constraint, r.client); err != nil {
+			glog.Errorf("Unable to create provider for constraint: %v", err)
```
Why you're swallowing the error here?

soltysh replied:
@deads2k quick question, is swallowing errors here ok, similarly to how it's done in [here](https://github.com/openshift/origin/blob/f0d3f88deb2083b0e86ad69069062ca38357748a/pkg/security/admission/admission.go#L144-L154) ?

soltysh commented on pkg/security/registry/podsecuritypolicyselfsubjectreview/rest.go:
```diff
@@ -0,0 +1,90 @@
+		}
+		filled, err := podsecuritypolicysubjectreview.FillPodSecurityPolicySubjectReviewStatus(&pspssr.Status, provider, pspssr.Spec.Template.Spec, constraint)
+		if err != nil {
+			glog.Errorf("unable to fill PodSecurityPolicySelfSubjectReview from constraint %v", err)
```
And here?

deads2k replied:
follow the same rules as admission.

soltysh commented on pkg/security/registry/podsecuritypolicyselfsubjectreview/rest_test.go:
```diff
@@ -0,0 +1,106 @@
+	oscc "github.com/openshift/origin/pkg/security/scc"
+)
+
+func validPodTemplateSpec() kapi.PodTemplateSpec {
```
Same problem here with tests, not covered errors paths.

soltysh commented on pkg/security/registry/podsecuritypolicysubjectreview/rest.go:
```diff
@@ -0,0 +1,115 @@
+		)
+		if provider, namespace, err = oscc.CreateProviderFromConstraint(ns, namespace, constraint, r.client); err != nil {
+			glog.Errorf("Unable to create provider for constraint: %v", err)
+			continue
```
Again, swallowing errors.

soltysh commented on pkg/security/registry/podsecuritypolicysubjectreview/rest.go:
```diff
@@ -0,0 +1,115 @@
+		filled, err := FillPodSecurityPolicySubjectReviewStatus(&pspsr.Status, provider, pspsr.Spec.Template.Spec, constraint)
+		if err != nil {
+			glog.Errorf("unable to fill PodSecurityPolicySubjectReviewStatus from constraint %v", err)
+			continue
```
And again.

soltysh commented on pkg/security/registry/podsecuritypolicysubjectreview/rest.go:
```diff
@@ -0,0 +1,115 @@
+	return pspsr, nil
+}
+
+// FillPodSecurityPolicySubjectReviewStatus fills PodSecurityPolicySubjectReviewStatus
```
This comment isn't much helpful. Can you describe what it fills? SCC, something else?

soltysh commented on pkg/security/registry/podsecuritypolicysubjectreview/rest.go:
```diff
@@ -0,0 +1,115 @@
+	}
+	if errs := oscc.AssignSecurityContext(provider, pod, field.NewPath(fmt.Sprintf("provider %s: ", provider.GetSCCName()))); len(errs) > 0 {
+		glog.Errorf("unable to assign SecurityContextConstraints provider: %v", errs)
+		//TODO: fills s.Reason
```
How about just putting the same text the error has into `s.Reason` instead of leaving this TODO?

soltysh commented on pkg/security/registry/podsecuritypolicysubjectreview/rest.go:
```diff
@@ -0,0 +1,115 @@
+	}
+	ref, err := kapi.GetReference(constraint)
+	if err != nil {
+		//TODO: fills s.Reason
```
Likewise...

soltysh commented on pkg/security/registry/podsecuritypolicysubjectreview/rest_test.go:
```diff
@@ -0,0 +1,148 @@
+	}
+}
+
+func TestAllowed(t *testing.T) {
```
Again, error paths should be tested.

soltysh commented:
It would be also nice, if you could split your PR into two: your changes and generated changes (iow. swagger-related stuff). It's easier for a reviewer to focus on the changes, only.

mfojtik commented:
@soltysh you mean splitting the commits right?

soltysh commented:
> @soltysh you mean splitting the commits right?

yup

sdminonne commented:
@soltysh 

- added test as requested
- modified comment
- added Reason string as requested
- as discussed via IRC error are not really swallowed, just trying to mimic security admission 
PTAL

soltysh commented:
With https://github.com/sdminonne/origin/pull/1 in I'll approve the changes, I want to see a clean test run, first.

sdminonne commented:
@soltysh thanks

sdminonne commented:
[test]

deads2k commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -165,6 +167,8 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
 					"selfsubjectrulesreviews", "subjectaccessreviews").RuleOrDie(),
 				authorizationapi.NewRule("create").Groups("authentication.k8s.io").Resources("tokenreviews").RuleOrDie(),
+				// permissions to check PSP, these creates are non-mutating
+				authorizationapi.NewRule("create").Groups(securityGroup).Resources("podsecuritypolicysubjectreviews", "podsecuritypolicyselfsubjectreviews", "podsecuritypolicyreviews").RuleOrDie(),
```
seems like a project admin ought to be able to use these too.  Already done?

deads2k commented on pkg/security/registry/podsecuritypolicyreview/rest.go:
```diff
@@ -0,0 +1,130 @@
+	if len(psprSpec.ServiceAccountNames) > 0 {
+		errs := []error{}
+		for _, saName := range psprSpec.ServiceAccountNames {
+			sa, err := client.Core().ServiceAccounts(namespace).Get(saName)
```
Do we already have a cache of service accounts in the API server to manage pod admission (betting yes).  If so, re-use that cache.

deads2k replied:
If its not already in a shared informer factory, then an issue with a link to here will do.

sdminonne replied:
@deads2k couldn't find. Opened https://github.com/openshift/origin/issues/11159

deads2k commented on pkg/security/registry/podsecuritypolicyreview/rest.go:
```diff
@@ -0,0 +1,130 @@
+		return nil, kapierrors.NewBadRequest(fmt.Sprintf("not a PodSecurityPolicyReview: %#v", obj))
+	}
+	if errs := securityvalidation.ValidatePodSecurityPolicyReview(pspr); len(errs) > 0 {
+		return nil, kapierrors.NewInvalid(securityapi.Kind(pspr.Kind), "", errs)
```
I think kind will be empty.  Hard coding the string is fine.

deads2k commented on pkg/security/registry/podsecuritypolicyreview/rest.go:
```diff
@@ -0,0 +1,130 @@
+
+	}
+	if len(errs) > 0 {
+		return nil, kerrors.NewAggregate(errs)
```
This needs to be wrapped in some kind of status error.

deads2k commented:
I think we want admission to collapse onto the same review mechanism at some point, but this would be a good start.

soltysh commented:
@deads2k if there are no objections I'd like to merge this on green tests.

sdminonne commented:
[test]

soltysh commented:
[merge]

soltysh commented:
flake #10327

re-[merge]

sdminonne commented:
@soltysh thanks!
